tools/nightly_checkout.py

```diff
@@ -0,0 +1,252 @@
+    for pkg in link:
+        url = URL_FORMAT.format(**pkg)
+        if pkg["name"] == "pytorch":
+            pytorch = url
+            platform = pkg["platform"]
+        else:
+            deps.append(url)
+    return deps, pytorch, platform
+
+
+@timed("Installing dependencies")
+def deps_install(deps):
+    """Install dependencies to deps environment"""
+    # first remove previous env
+    cmd = ["conda", "env", "remove", "--yes", "--name", "pytorch-deps"]
+    p = subprocess.run(cmd, check=True)
+    # install new deps
+    cmd = ["conda", "create", "--yes", "--no-deps", "--name", "pytorch-deps"] + deps
+    p = subprocess.run(cmd, check=True)
+
+
+@timed("Installing pytorch nightly binaries")
+def pytorch_install(url):
+    """"Install pytorch into a temporary directory"""
+    pytdir = tempfile.TemporaryDirectory()
+    cmd = ["conda", "create", "--yes", "--no-deps", "--prefix", pytdir.name, url]
+    p = subprocess.run(cmd, check=True)
+    return pytdir
+
+
+def _site_packages(pytdir, platform):
+    if platform.startswith("win"):
+        os.path.join(pytdir.name, "Lib", "site-packages")
+    else:
+        template = os.path.join(pytdir.name, "lib", "python*.*", "site-packages")
+        spdir = glob.glob(template)[0]
+    return spdir
+
+
+def _ensure_commit(git_sha1):
+    """Make sure that we actually have the commit locally"""
+    cmd = ["git", "cat-file", "-e", git_sha1 + "^{commit}"]
+    p = subprocess.run(cmd, capture_output=True, check=False)
+    if p.returncode == 0:
+        # we have the commit locally
+        return
+    # we don't have the commit, must fetch
+    cmd = ["git", "fetch", "https://github.com/pytorch/pytorch.git", git_sha1]
+    p = subprocess.run(cmd, check=True)
+
+
+@timed("Checking out nightly PyTorch")
+def checkout_nightly_version(spdir):
+    """Get's the nightly version and then checks it out."""
+    # first get the git version from the installed module
+    version_fname = os.path.join(spdir, "torch", "version.py")
+    with open(version_fname) as f:
+        lines = f.read().splitlines()
+    for line in lines:
+        if not line.startswith('git_version'):
+            continue
+        git_version = literal_eval(line.partition("=")[2].strip())
+        break
+    else:
+        raise RuntimeError(f"Could not find git_version in {version_fname}")
+    print(f"Found released git version {git_version}")
+    # now cross refernce with nightly version
+    _ensure_commit(git_version)
+    cmd = ["git", "show", "--no-patch", "--format=%s", git_version]
+    p = subprocess.run(cmd, capture_output=True, check=True, text=True)
+    m = SHA1_RE.search(p.stdout)
+    if m is None:
+        raise RuntimeError(f"Could not find nightly release in git history:\n  {p.stdout}")
+    nightly_version = m.group(1)
+    print(f"Found nightly release version {nightly_version}")
+    # now checkout nightly version
+    _ensure_commit(nightly_version)
+    cmd = ["git", "checkout", nightly_version]
```

Comment:
Clobbering whatever local changes people had when you do this doesn't sound great.  Something more clever is going to need to happen here.

Reply:
OK, I think we should require a branch name that they will checkout then.